samples/guide/src/main/java/com/squareup/okhttp/recipes/Progress.java

```diff
@@ -0,0 +1,119 @@
+ */
+package com.squareup.okhttp.recipes;
+
+import com.squareup.okhttp.Interceptor;
+import com.squareup.okhttp.MediaType;
+import com.squareup.okhttp.OkHttpClient;
+import com.squareup.okhttp.Response;
+import com.squareup.okhttp.ResponseBody;
+import com.squareup.okhttp.Request;
+import java.io.IOException;
+import okio.Buffer;
+import okio.BufferedSource;
+import okio.ForwardingSource;
+import okio.Okio;
+import okio.Source;
+
+public final class Progress {
+
+  private final OkHttpClient client = new OkHttpClient();
+
+  public void run() throws Exception {
+    Request request = new Request.Builder()
+        .url("https://publicobject.com/helloworld.txt")
+        .build();
+
+    client.networkInterceptors().add(new Interceptor() {
+      @Override public Response intercept(Chain chain) throws IOException {
+        Response originalResponse = chain.proceed(chain.request());
+        return originalResponse.newBuilder()
+            .body(new ProgressResponseBody(
+                originalResponse.body().contentLength(),
+                originalResponse.body().contentType(),
+                new ProgressListener(),
+                originalResponse.body().source()))
+            .build();
+        }
+    });
+
+    Response response = client.newCall(request).execute();
+    if (!response.isSuccessful()) throw new IOException("Unexpected code " + response);
+
+    System.out.println(response.body().string());
+  }
+
+  public static void main(String... args) throws Exception {
+    new Progress().run();
+  }
+
+  private static class ProgressResponseBody extends ResponseBody {
+
+    private long limit;
+    private MediaType contentType;
+    private ProgressListener progressListener;
+    private Source source;
+
+    public ProgressResponseBody(final long limit, final MediaType contentType,
+        final ProgressListener progressListener, Source source) {
+      this.limit = limit;
+      this.contentType = contentType;
+      this.progressListener = progressListener;
+      this.source = source;
+    }
+
+    @Override public MediaType contentType() {
+      return contentType;
+    }
+
+    @Override public long contentLength() {
+      return limit;
+    }
+
+    @Override public BufferedSource source() {
+      return Okio.buffer(source(source));
+    }
+
+    private Source source(Source source) {
+      return new ForwardingSource(source) {
+        long totalBytesRead = 0L;
+        @Override public long read(Buffer sink, long byteCount) throws IOException {
+          long bytesRead = super.read(sink, byteCount);
+          // read() returns the number of bytes read, or -1 if this source is exhausted.
+          totalBytesRead += bytesRead != -1 ? bytesRead : 0;
+          progressListener.update(totalBytesRead, limit, false);
+          return bytesRead;
+        }
+
+        @Override public void close() throws IOException {
+          super.close();
+          progressListener.update(totalBytesRead, limit, true);
```

Comment:
And if you drop the bytesRead special case above, you don't need to override `close()` at all? Or do you want to do that in case the request gets canceled? I suspect it's simpler to just not override close and not make promises.